File: src/competitve practice/leetcode-30day-challenge/week4/SubArraySumEqualsk.java

class Solution {
    public int subarraySum(int[] nums, int k) {
        long sumArr[] = new long[nums.length];
        sumArr[0] = nums[0];

        for (int i = 1; i < nums.length; i++) {
            sumArr[i] = sumArr[i - 1] + nums[i];
        }

        int count = 0;
        HashMap<Long, Integer> sumMap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (sumArr[i] == k) {
                count++;
            }
            if (sumMap.containsKey(sumArr[i] - k)) {
                count += sumMap.get(sumArr[i] - k);
            }

            sumMap.put(sumArr[i], sumMap.getOrDefault(sumArr[i], 0) + 1);

        }
        return count;
    }
}

class Solution2 {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> sumMap = new HashMap<Integer, Integer>();
        sumMap.put(0, 1);
        int count = 0;
        int sum = 0;

        for (int i = 0; i < nums.length; i++) {
            sum = sum + nums[i];
            if (sumMap.containsKey(sum - k)) {
                count = count + sumMap.get(sum - k);
            }
            sumMap.put(sum, sumMap.getOrDefault(sum, 0) + 1);
        }
        return count;
    }
}

File: LeetCode/leetcode-30day-challenge/week4/C++/SubArraySumEqualsk.cpp
#include <bits/stdc++.h>
using namespace std;

//Type: Arrays, Maps

//Approach: Maintain a map to store all the frequencies of unique sums
//          till now. Now, saya t a current index i, the totalSum of elements
//          is S. So, we will have to find out whether a sum of (S-k) occured
//          previously. If yes, add the number of times we found it.
//          Increment the count of current sum in the map.
//Complexity: O(N) time and O(N) space

class Solution
{
public:
    int subarraySum(vector<int> &nums, int k)
    {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        unordered_map<int, int> sumMap;
        sumMap[0] = 1;
        int count = 0;
        int sum = 0;

        for (int i = 0; i < nums.size(); i++)
        {
            sum = sum + nums[i];
            if (sumMap.count(sum - k))
            {
                count += sumMap[sum - k];
            }
            sumMap[sum] = sumMap[sum] + 1;
        }
        return count;
    }
};